src/app/layout/account-switcher.component.ts

```diff
@@ -0,0 +1,77 @@
+import {
+    animate,
+    state,
+    style,
+    transition,
+    trigger,
+  } from '@angular/animations';
+import { Component, OnInit } from '@angular/core';
+import { Router } from '@angular/router';
+
+import { MessagingService } from 'jslib-common/abstractions/messaging.service';
+import { StateService } from 'jslib-common/abstractions/state.service';
+import { VaultTimeoutService } from 'jslib-common/abstractions/vaultTimeout.service';
+import { AuthenticationStatus } from 'jslib-common/enums/authenticationStatus';
+
+import { Account } from 'jslib-common/models/domain/account';
+
+@Component({
+    selector: 'app-account-switcher',
+    templateUrl: 'account-switcher.component.html',
+    animations: [
+        trigger('transformPanel', [
+            state('void', style({
+              opacity: 0,
+            })),
+            transition('void => open',  animate('100ms linear', style({
+              opacity: 1,
+            }))),
+            transition('* => void', animate('100ms linear', style({opacity: 0}))),
+        ]),
+    ],
+})
+export class AccountSwitcherComponent implements OnInit {
+    isOpen: boolean = false;
+    accounts: { [userId: string]: Account } = {};
+    activeAccountEmail: string;
+
+    get showSwitcher() {
+        return this.accounts != null && Object.keys(this.accounts).length > 0;
+    }
+
+    constructor(private stateService: StateService, private vaultTimeoutService: VaultTimeoutService,
+        private messagingService: MessagingService, private router: Router) {}
+
+    async ngOnInit(): Promise<void> {
+        this.stateService.accounts.subscribe(async accounts => {
+            for (const userId in accounts) {
+                if (userId === await this.stateService.getUserId()) {
+                    accounts[userId].profile.authenticationStatus = AuthenticationStatus.Active;
+                } else {
+                    accounts[userId].profile.authenticationStatus = await this.vaultTimeoutService.isLocked(userId) ?
+                        AuthenticationStatus.Locked :
+                        AuthenticationStatus.Unlocked;
+                }
+            }
+
+            this.accounts = accounts;
+            this.activeAccountEmail = await this.stateService.getEmail();
+        });
+    }
+
+    toggle() {
+        this.isOpen = !this.isOpen;
+    }
+
+    async switch(userId: string) {
+        await this.stateService.setActiveUser(userId);
+        const locked = await this.vaultTimeoutService.isLocked(userId);
+        if (locked) {
+            this.messagingService.send('locked', { userId: userId });
+        } else {
+            this.messagingService.send('unlocked');
+            this.messagingService.send('syncVault');
+            this.router.navigate(['vault']);
+        }
```

Comment:
It's not exactly duplicated. Logout checks for any other accounts to switch to, and automatically switches if it find one or redirects to the login page if it doesn't find one. This just checks for if an account is locked or unlocked on switch and handles accordingly. In your comment above you mentioned that we should subscribe to account switched and handle accordingly, but this isn't really an account switch in the same sense as a manual one. I could see combining some of this into a new message and bundling them that way, but we can't run the same exact logic over both this scenario and the logout scenerio - and handling this with a subscription to stateService.activeAccount has other consequences for operations like initial boot. I'm going to think about this some more, because you're right it isn't very elegant, but there is more to consider.

Reply:
Fair enough. Could we have a tech debt item for this? 😁. In general I prefer having as lightweight components as possible with the logic living elsewhere. (Components are the UI layer)

Reply:
https://app.asana.com/0/1201461753012676/1201521717039124